### PR TITLE
COMMERCE-9387 Dynamic field fragment

### PR DESCRIPTION
**Pull request by smottal**
https://issues.liferay.com/browse/COMMERCE-9387
https://issues.liferay.com/browse/COMMERCE-9386
https://issues.liferay.com/browse/COMMERCE-9388
https://issues.liferay.com/browse/COMMERCE-9389

### CODE REVIEW

**Comment by smottal** [comment]
@andrea-ale-sbarra @alessiorendina for review

File: modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import * as Events from 'commerce-frontend-js/utilities/eventsDefinitions';
+
+const deep_value = (obj, path, defaultValue) => {
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js:17]: just check the function name pattern I'm not sure about the _ inside the function name

**Review by andrea-ale-sbarra — APPROVED** [review]

**Comment by andrea-ale-sbarra** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import * as Events from 'commerce-frontend-js/utilities/eventsDefinitions';
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js:15]: maybe this * is not right

Comment (andrea-ale-sbarra) [reply modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js:15]: try to import only what you use

Comment (smottal) [reply modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/fragment/renderer/dynamic_field/js/CPInstanceChangeHandler.js:15]: It's the same inside modules/apps/commerce/commerce-product-content-web/src/main/resources/META-INF/resources/info/item/renderer/availability_label/js/AvailabilityCPInstanceChangeHandler.js

**Comment by smottal** [comment]
ci:test:sf

**Comment by smottal** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-product-content-web/src/main/resources/com/liferay/fragment/renderer/dynamic_field/dependencies/configuration.json [hunk 3]
```diff
@@ -0,0 +1,133 @@
+{
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-content-web/src/main/resources/com/liferay/fragment/renderer/dynamic_field/dependencies/configuration.json:1]: Move this file under com/liferay/commerce/product/content/web/internal/fragment/renderer/dynamic_field/dependencies

File: modules/apps/commerce/commerce-product-content-web/src/main/resources/com/liferay/fragment/renderer/dynamic_field/dependencies/configuration.json [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+								"label": "Inventory",
+								"value": "availability.stockQuantity"
+							},
+							{
+								"label": "Manufacturer Part Number",
+								"value": "manufacturerPartNumber"
+							},
+							{
+								"label": "SKU",
+								"value": "sku"
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-content-web/src/main/resources/com/liferay/fragment/renderer/dynamic_field/dependencies/configuration.json:26]: Pass localized key to the label props

File: modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+			requestDispatcher.include(httpServletRequest, httpServletResponse);
+		}
+		catch (Exception exception) {
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private String _getConfigurationValue(
+		FragmentEntryLink fragmentEntryLink, String type) {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java:209]: type -> name

File: modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+		httpServletRequest.setAttribute(
+			"liferay-commerce:dynamic-field:field", field);
+		httpServletRequest.setAttribute(
+			"liferay-commerce:dynamic-field:label", label);
+		httpServletRequest.setAttribute(
+			"liferay-commerce:dynamic-field:labelElementType",
+			labelElementType);
+		httpServletRequest.setAttribute(
+			"liferay-commerce:dynamic-field:valueElementType",
+			valueElementType);
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java:148]: Remove variables and inline them

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 7]
```diff
@@ -9609,6 +9611,7 @@ mapped-products=Mapped Products
 mapped-url=Mapped URL
 mapping=Mapping
 mapping-description[oauth2]=Map key into value using "key=value" syntax. To group multiple keys and values together, use "key1,key2,key3=value1,value2,value3".
+mapping-options=Mapping Options
```
Comment (alessiorendina) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:9614]: Let's mimic the existing configs and remove this label

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@smottal please check my comments, thanks!

**Comment by smottal** [comment]
ci:test:commerce

File: modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+			JSONArray jsonArray3 = jsonObject2.getJSONArray("validValues");
+
+			for (Object validValueObject : jsonArray3) {
+				JSONObject jsonObject3 = (JSONObject)validValueObject;
+
+				String value = jsonObject3.getString("value");
+
+				if (value.equals(field)) {
+					return jsonObject3.getString("label");
+				}
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java:297]: Is there any other way to do this?

Comment (smottal) [reply modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java:297]: Unfortunately I didn't found another way to get the "label" of the selected item from the configuration. I have to parse the json to find the selected option by the "selected value"

Comment (smottal) [reply modules/apps/commerce/commerce-product-content-web/src/main/java/com/liferay/commerce/product/content/web/internal/fragment/renderer/DynamicFieldFragmentRenderer.java]: Done

**Comment by smottal** [comment]
ci:test:sf

**Comment by smottal** [comment]
ci:test:sf

**Comment by alessiorendina** [comment]
ci:test:relevant

**Review by alessiorendina — APPROVED** [review]

**Comment by lorenzolapescara** [comment]
Re-sent at #3553

**Comment by alessiorendina** [comment]
Forwarded here: https://github.com/liferay-commerce/liferay-portal/pull/3553

**Pull request closed without merging**